File: python/unique_anagram/ana.py

def factorial(n):
    fact = 1
    for i in range(1, n+1):
        fact = fact * i
    return fact

def anagram(string, output, output_index):

    length = len(string)
    for i in range(length):

        pick = string[i]
        remainder = string[0:i] + string[i+1:length]

        for j in range(0, factorial(length-1)):
            output[ output_index + j] += pick

        anagram(remainder, output, output_index)

        output_index += factorial(length-1);

def deduplicate(output, number):
    count = 0
    for i in range(number):
        if output[i] != "":
            count += 1
            for j in range(i+1, number):
                if output[i] == output[j]:
                    output[j] = ""
    return count

def main():
    a = "pineapple"

    number = len(a)
    print("Number of letters: {}".format(number))

    anagrams = factorial(number)
    print("Number of anagrams: {}".format(anagrams))

    output = [""] * anagrams

    anagram(a, output, 0)

    print("First anagram: {}".format(output[0]))
    print("Last anagram: {}".format(output[anagrams-1]))
    count = deduplicate(output, anagrams)
    print("Number of unique anagrams: {}".format(count))

if __name__ == "__main__":
    main()

File: go/unique_anagram/ana.go
package main

import "fmt"

func factorial(n int) int {
    var fact int = 1
    for i:=1; i < n+1; i+=1 {
        fact = fact * i
	}
    return fact
}

func anagram(astring string, output[] string, output_index int){
	var length int = len(astring)
    for i:=0; i<length; i+=1 {

        a := []byte(astring)
        var pick byte = a[i]
        var remainder[] byte = append(a[0:i], a[i+1:length]...)
		
        for j:=0; j<factorial(length - 1); j+=1 {
            output[ output_index + j] += string(pick)
		}
		
        anagram(string(remainder), output, output_index)

        output_index += factorial(length - 1)
	}
}

func deduplicate(output[] string, number int) int {
    var count int = 0
	for i, ana1 := range output {
        if ana1 != "" {
            count++
            for j := i+1; j<number; j+=1 {
                if ana1 == output[j] {
                    output[j] = ""
				}
            }
        }
    }
    return count
}

func main() {
    const a string = "pineapple"

    const number int = len(a)
    fmt.Printf("Number of letters: %d\n",number)

    var anagrams int = factorial(number)
    fmt.Printf("Number of anagrams: %d\n",anagrams)

	output := make([]string, anagrams)
	
    anagram(a, output, 0);

    fmt.Printf("First anagram: %s\n",output[0]);
    fmt.Printf("Last anagram: %s\n",output[anagrams-1]);
    var count int = deduplicate(output, anagrams);
    fmt.Printf("Number of unique anagrams: %d\n",count);
}